### PR TITLE
feat(flags): modify the order and priority when loading the configs

### PR DESCRIPTION
**Pull request by YoGhurt111**

### CODE REVIEW

File: cmd/flags/common.go [hunk 1]
```diff
@@ -128,6 +128,12 @@ var (
 		Category: commonCategory,
 		Value:    12 * time.Second,
 	}
+	// ConfigFile Path to a config file
+	ConfigFile = &cli.StringFlag{
+		Name:     "config.file",
```
Comment (davidtaikocha) [inline cmd/flags/common.go:133]: Let's use `config` here (and change the flag description), to keep the same as what go-ethereum is using:

image

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+	"unicode"
+
+	"github.com/naoina/toml"
+)
+
+var tomlSettings = toml.Config{
```
Comment (davidtaikocha) [inline cmd/utils/flags.go:14]:
```suggestion
// These settings ensure that TOML keys use the same names as Go struct fields.
var tomlSettings = toml.Config{
```

File: cmd/utils/flags.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	MissingField: func(rt reflect.Type, field string) error {
+		var link string
+		if unicode.IsUpper(rune(rt.Name()[0])) && rt.PkgPath() != "main" {
+			link = fmt.Sprintf(", see https://godoc.org/%s#%s for available fields", rt.PkgPath(), rt.Name())
+		}
+		return fmt.Errorf("field '%s' is not defined in %s%s", field, rt.String(), link)
+	},
+}
+
+func LoadConfigFile(file string, cfg interface{}) error {
```
Comment (davidtaikocha) [inline cmd/utils/flags.go:30]: lets add a comment for this function?

File: driver/config.go [hunk 4]
```diff
@@ -28,51 +29,102 @@ type Config struct {
 // NewConfigFromCliContext creates a new config instance from
 // the command line inputs.
 func NewConfigFromCliContext(c *cli.Context) (*Config, error) {
-	jwtSecret, err := jwt.ParseSecretFromFile(c.String(flags.JWTSecret.Name))
-	if err != nil {
-		return nil, fmt.Errorf("invalid JWT secret file: %w", err)
+	// Defaults config
+	cfg := Config{
+		ClientConfig: &rpc.ClientConfig{
+			// TODO: To be confirmed whether taiko addresses in L1 or L2 are constant
```
Comment (davidtaikocha) [inline driver/config.go:35]: Yes they are proxy addresses, so won't change in future.

File: driver/config.go [hunk 5]
```diff
@@ -28,51 +29,102 @@ type Config struct {
...
-	)
+	// Load config file
+	if file := c.String(flags.ConfigFile.Name); file != "" {
+		if err := utils.LoadConfigFile(file, &cfg); err != nil {
+			return nil, fmt.Errorf("%w", err)
+		}
+	}
+	// Apply flag value
+	err := ApplyFlagValue(c, &cfg)
+	if err != nil {
```
Comment (davidtaikocha) [inline driver/config.go:55]:
```suggestion
	if err := ApplyFlagValue(c, &cfg); err != nil {
```

If the function only return an error, lets use `if err := ...; err !=nil {`

File: cmd/utils/flags.go (hunk 3 above)
Comment (YoGhurt111) [reply cmd/utils/flags.go:30]: ok

File: cmd/flags/common.go (hunk 1 above)
Comment (YoGhurt111) [reply cmd/flags/common.go:133]: ok

File: driver/config.go [hunk 6]
```diff
@@ -28,51 +29,100 @@ type Config struct {
...
 func NewConfigFromCliContext(c *cli.Context) (*Config, error) {
-	jwtSecret, err := jwt.ParseSecretFromFile(c.String(flags.JWTSecret.Name))
-	if err != nil {
-		return nil, fmt.Errorf("invalid JWT secret file: %w", err)
+	// Defaults config
+	cfg := Config{
+		ClientConfig: &rpc.ClientConfig{
+			Timeout: 12 * time.Second,
+		},
+		RetryInterval:         backoff.DefaultMaxInterval,
```
Comment (davidtaikocha) [inline driver/config.go:37]: We should define the default value in flag definition, and use `flags.BackOffRetryInterval.Value` here

File: cmd/flags/common.go [hunk 7]
```diff
@@ -128,6 +128,12 @@ var (
 		Category: commonCategory,
 		Value:    12 * time.Second,
 	}
+	// ConfigFile TOML configuration file path
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply cmd/flags/common.go:131]: We already have the flag description below, no need for this comment here.

File: driver/config.go [hunk 8]
```diff
@@ -28,51 +28,100 @@ type Config struct {
...
+		RPCTimeout:            flags.RPCTimeout.Value,
+		MaxExponent:           flags.MaxExponent.Value,
 	}
 
-	var (
-		p2pSyncVerifiedBlocks = c.Bool(flags.P2PSyncVerifiedBlocks.Name)
-		l2CheckPoint          = c.String(flags.CheckPointSyncURL.Name)
-	)
+	// Load config file
+	if file := c.String(flags.ConfigFile.Name); file != "" {
```
Comment (davidtaikocha) [inline driver/config.go:44]: Let's simply use `c.IsSet(...)`?

**Pull request closed without merging**